### PR TITLE
fix edge case asymmetric number of unaligned reads

### PR DESCRIPTION
**Pull request by nlhepler**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -206,6 +206,12 @@ impl StarAligner {
         rec
     }
 
+    fn empty_records(name: &[u8], read: &[u8], qual: &[u8], n: usize) -> Vec<bam::Record> {
```
Comment (pmarks) [inline src/lib.rs:209]: Looking at this, do we have a few other issues?
I think the empty records should have:
- 0x4 / 0x8 flags set to declare whether they are R1 or R1.
- Exactly one of the records must be primary (flags & 0x900 == 0), and should probably form a pair with the primary record that was returned by STAR.
I think we might confuse bamtofastq without those properties.

Comment (nlhepler) [reply src/lib.rs:209]: Good call.

I have fixed this up below (when we know whether first or last segment in the template).

File: src/lib.rs [hunk 2]
```diff
@@ -236,12 +242,26 @@ impl StarAligner {
...
         if read1.len() == 0 {
-            let recs1 = vec![Self::empty_record(name, read1, qual1)];
-            let recs2 = self.align_read(name, read2, qual2);
+            let mut recs2 = self.align_read(name, read2, qual2);
+            let mut recs1 = Self::empty_records(name, read1, qual1, recs2.len());
+            for (rec1, rec2) in recs1.iter_mut().zip(&mut recs2) {
+                if rec2.is_unmapped() {
+                    rec1.set_mate_unmapped();
+                }
+                if rec2.is_supplementary() {
```
Comment (pmarks) [inline src/lib.rs:251]: i think you also want the same thing for `is_secondary()` as well

File: src/lib.rs [hunk 3]
```diff
@@ -236,12 +242,26 @@ impl StarAligner {
...
+            }
             return (recs1, recs2);
         } else if read2.len() == 0 {
-            let recs1 = self.align_read(name, read1, qual1);
-            let recs2 = vec![Self::empty_record(name, read2, qual2)];
+            let mut recs1 = self.align_read(name, read1, qual1);
+            let mut recs2 = Self::empty_records(name, read2, qual2, recs1.len());
+            for (rec1, rec2) in recs1.iter_mut().zip(&mut recs2) {
+                rec1.set_mate_unmapped();
+                if rec1.is_supplementary() {
```
Comment (pmarks) [inline src/lib.rs:261]: here as well

**Comment by nlhepler** [comment]
Okay, _this_ should cover it. Referring to [SAM/BAM spec](https://samtools.github.io/hts-specs/SAMv1.pdf)
> 0x1 template having multiple segments in sequencing (`set_paired`)
> 0x8 _next_ segment in the template unmapped (emphasis mine, `set_mate_unmapped`)
> 0x40 the first segment in the template (`set_first_in_template`)
> 0x80 the last segment in the template (`set_last_in_template`)
> 0x100 secondary alignment (`set_secondary`)
> 0x800 supplementary alignment (`set_supplementary`)

**Comment by nlhepler** [comment]
Yeah. If STAR does that, then we'll just use the first one we see. Which at least doesn't crash things.

**Pull request merged**